<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>A group of helper functions for manipulating and cleaning a
MultiAssayExperiment — MultiAssayExperiment-helpers • MultiAssayExperiment</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="A group of helper functions for manipulating and cleaning a
MultiAssayExperiment — MultiAssayExperiment-helpers" />
<meta property="og:description" content="A set of helper functions were created to help clean and
manipulate a MultiAssayExperiment object. intersectRows also works
for ExperimentList objects.

complete.cases: Returns a logical vector corresponding to 'colData'
    rows that have data across all experiments
isEmpty: Returns a logical TRUE value for zero length
    MultiAssayExperiment objects
intersectRows: Takes all common rows across experiments,
    excludes experiments with empty rownames
intersectColumns: A wrapper for complete.cases to return a
    MultiAssayExperiment with only those biological units that have
    measurements across all experiments
replicated: A function that identifies multiple samples that
    originate from a single biological unit within each assay
anyReplicated: Displays which assays have replicate measurements
mergeReplicates: A function that combines replicated / repeated
    measurements across all experiments and is guided by the replicated
    return value
longFormat: A MultiAssayExperiment method that
    returns a small and skinny DataFrame. The colDataCols
    arguments allows the user to append colData columns to the data.
wideFormat: A function to return a wide DataFrame where
    each row represents an observation. Optional colDataCols can be
    added when using a MultiAssayExperiment.
hasRowRanges: A function that identifies ExperimentList elements
    that have a rowRanges method
getWithColData: A convenience function for extracting an assay
    and associated colData

" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MultiAssayExperiment</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.12.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Quick Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/QuickStartMultiAssay.html">Quick Start Guide</a>
    </li>
    <li>
      <a href="../articles/MultiAssayExperiment.html">Getting Started</a>
    </li>
    <li>
      <a href="../articles/UsingHDF5Array.html">Using DelayedMatrix</a>
    </li>
    <li>
      <a href="../articles/extras/MAE-curatedMetagenomicData.html">MultiAssayExperiment in Practice</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">
    <span class="fa fa-newspaper-o"></span>
     
    NEWS
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/waldronlab/MultiAssayExperiment">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>A group of helper functions for manipulating and cleaning a
MultiAssayExperiment</h1>
    <small class="dont-index">Source: <a href='https://github.com/waldronlab/MultiAssayExperiment/blob/master/R/MultiAssayExperiment-helpers.R'><code>R/MultiAssayExperiment-helpers.R</code></a></small>
    <div class="hidden name"><code>MultiAssayExperiment-helpers.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>A set of helper functions were created to help clean and
manipulate a MultiAssayExperiment object. <code>intersectRows</code> also works
for <code>ExperimentList</code> objects.</p>
<ul>
<li><p>complete.cases: Returns a logical vector corresponding to 'colData'
    rows that have data across all experiments</p></li>
<li><p>isEmpty: Returns a logical <code>TRUE</code> value for zero length
    <code>MultiAssayExperiment</code> objects</p></li>
<li><p>intersectRows: Takes all common rows across experiments,
    excludes experiments with empty rownames</p></li>
<li><p>intersectColumns: A wrapper for <code>complete.cases</code> to return a
    <code>MultiAssayExperiment</code> with only those biological units that have
    measurements across all experiments</p></li>
<li><p>replicated: A function that identifies multiple samples that
    originate from a single biological unit within each assay</p></li>
<li><p>anyReplicated: Displays which assays have replicate measurements</p></li>
<li><p>mergeReplicates: A function that combines replicated / repeated
    measurements across all experiments and is guided by the replicated
    return value</p></li>
<li><p>longFormat: A <code>MultiAssayExperiment</code> method that
    returns a small and skinny <a href='https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html'>DataFrame</a>. The <code>colDataCols</code>
    arguments allows the user to append <code>colData</code> columns to the data.</p></li>
<li><p>wideFormat: A function to return a wide <a href='https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html'>DataFrame</a> where
    each row represents an observation. Optional <code>colDataCols</code> can be
    added when using a <code>MultiAssayExperiment</code>.</p></li>
<li><p>hasRowRanges: A function that identifies ExperimentList elements
    that have a rowRanges method</p></li>
<li><p>getWithColData: A convenience function for extracting an assay
    and associated colData</p></li>
</ul>

    </div>

    <pre class="usage"><span class='co'># S4 method for MultiAssayExperiment</span>
<span class='fu'><a href='https://rdrr.io/r/stats/complete.cases.html'>complete.cases</a></span>(<span class='no'>...</span>)

<span class='co'># S4 method for MultiAssayExperiment</span>
<span class='fu'>isEmpty</span>(<span class='no'>x</span>)

<span class='fu'>intersectRows</span>(<span class='no'>x</span>)

<span class='fu'>intersectColumns</span>(<span class='no'>x</span>)

<span class='fu'>replicated</span>(<span class='no'>x</span>)

<span class='co'># S4 method for MultiAssayExperiment</span>
<span class='fu'>replicated</span>(<span class='no'>x</span>)

<span class='fu'>anyReplicated</span>(<span class='no'>x</span>)

<span class='co'># S4 method for MultiAssayExperiment</span>
<span class='fu'>anyReplicated</span>(<span class='no'>x</span>)

<span class='fu'>mergeReplicates</span>(<span class='no'>x</span>, <span class='kw'>replicates</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span>(), <span class='kw'>simplify</span> <span class='kw'>=</span> <span class='kw pkg'>BiocGenerics</span><span class='kw ns'>::</span><span class='no'><a href='https://rdrr.io/pkg/BiocGenerics/man/mean.html'>mean</a></span>, <span class='no'>...</span>)

<span class='co'># S4 method for MultiAssayExperiment</span>
<span class='fu'>mergeReplicates</span>(
  <span class='no'>x</span>,
  <span class='kw'>replicates</span> <span class='kw'>=</span> <span class='fu'>replicated</span>(<span class='no'>x</span>),
  <span class='kw'>simplify</span> <span class='kw'>=</span> <span class='kw pkg'>BiocGenerics</span><span class='kw ns'>::</span><span class='no'><a href='https://rdrr.io/pkg/BiocGenerics/man/mean.html'>mean</a></span>,
  <span class='no'>...</span>
)

<span class='co'># S4 method for ANY</span>
<span class='fu'>mergeReplicates</span>(<span class='no'>x</span>, <span class='kw'>replicates</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span>(), <span class='kw'>simplify</span> <span class='kw'>=</span> <span class='kw pkg'>BiocGenerics</span><span class='kw ns'>::</span><span class='no'><a href='https://rdrr.io/pkg/BiocGenerics/man/mean.html'>mean</a></span>, <span class='no'>...</span>)

<span class='fu'>longFormat</span>(<span class='no'>object</span>, <span class='kw'>colDataCols</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>i</span> <span class='kw'>=</span> <span class='fl'>1L</span>)

<span class='fu'>wideFormat</span>(
  <span class='no'>object</span>,
  <span class='kw'>colDataCols</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>check.names</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>collapse</span> <span class='kw'>=</span> <span class='st'>"_"</span>,
  <span class='kw'>i</span> <span class='kw'>=</span> <span class='fl'>1L</span>
)

<span class='fu'>hasRowRanges</span>(<span class='no'>x</span>)

<span class='co'># S4 method for MultiAssayExperiment</span>
<span class='fu'>hasRowRanges</span>(<span class='no'>x</span>)

<span class='co'># S4 method for ExperimentList</span>
<span class='fu'>hasRowRanges</span>(<span class='no'>x</span>)

<span class='fu'>getWithColData</span>(<span class='no'>x</span>, <span class='no'>i</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>...</th>
      <td><p>Additional arguments. See details for more information.</p></td>
    </tr>
    <tr>
      <th>x</th>
      <td><p>A MultiAssayExperiment or ExperimentList</p></td>
    </tr>
    <tr>
      <th>replicates</th>
      <td><p>A list of LogicalLists
indicating multiple / duplicate entries for each biological unit per assay,
see <code>replicated</code> (default <code>replicated(x)</code>).</p></td>
    </tr>
    <tr>
      <th>simplify</th>
      <td><p>A function for merging repeat measurements in experiments
as indicated by the <code>replicated</code> method for <code>MultiAssayExperiment</code></p></td>
    </tr>
    <tr>
      <th>object</th>
      <td><p>Any supported class object</p></td>
    </tr>
    <tr>
      <th>colDataCols</th>
      <td><p>A <code>character</code>, <code>logical</code>, or <code>numeric</code>
index for <code>colData</code> columns to be included</p></td>
    </tr>
    <tr>
      <th>i</th>
      <td><p>The assay indicator for SummarizedExperiment
objects (default 1L)</p></td>
    </tr>
    <tr>
      <th>check.names</th>
      <td><p>(logical default TRUE) Column names of the output
<code>DataFrame</code> will be checked for syntactic validity and made unique,
if necessary</p></td>
    </tr>
    <tr>
      <th>collapse</th>
      <td><p>(character default "_") A single string delimiter for output
column names. In <code>wideFormat</code>, experiments and rownames (and when
replicate samples are present, colnames) are seperated by this delimiter</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The <code>replicated</code> function finds replicate measurements in each
assay and returns a list of LogicalLists.
Each element in a single LogicalList corresponds to a
biological or <em>primary</em> unit as in the <code>sampleMap</code>. Below is a
small graphic for one particular biological unit in one assay, where the
logical vector corresponds to the number of measurements/samples in the
assay:</p><pre>
 &amp;gt;      replicated(MultiAssayExperiment)
 (list str)       '-- $ AssayName
 (LogicalList str)      '-- [[ "Biological Unit" ]]
 Replicated if sum(...) &amp;gt; 1          '-- TRUE TRUE FALSE FALSE
</pre><p><code>anyReplicated</code> determines if any of the assays have at least one
replicate. <em>Note</em>. These methods are not available for the
<code>ExperimentList</code> class due to a missing <code>sampleMap</code> structure
(by design).</p>
<p>The <code>mergeReplicates</code> function is a house-keeping method
for a <code>MultiAssayExperiment</code> where only <code>complete.cases</code> are
returned, replicate measurements are averaged (by default), and columns are
aligned by the row order in <code>colData</code>. Additional arguments can be
passed on to the <code>simplify</code> function.</p>
<p>The <code>mergeReplicates</code> "ANY" method consolidates duplicate
measurements for rectangular data structures, returns object of the same
class (endomorphic). The ellipsis or <code>...</code> argument allows the
user to provide additional arguments to the <code>simplify</code> functional
argument.</p>
<p>The <code>longFormat</code> "ANY" class method, works with classes such as
ExpressionSet and SummarizedExperiment as well as <code>matrix</code>
to provide a consistent long and skinny <a href='https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html'>DataFrame</a>.</p>
<p>The <code>hasRowRanges</code> method identifies assays that support
a rowRanges method <em>and</em>
return a GRanges object.</p>
    <h2 class="hasAnchor" id="mergereplicates"><a class="anchor" href="#mergereplicates"></a>mergeReplicates</h2>

    

<p>The <code>mergeReplicates</code> function makes use of the output from
<code>replicated</code> which will point out the duplicate measurements by
biological unit in the <code>MultiAssayExperiment</code>. This function will return
a <code>MultiAssayExperiment</code> with merged replicates. Additional arguments
can be provided to the simplify argument via the ellipsis (...).</p>
    <h2 class="hasAnchor" id="longformat"><a class="anchor" href="#longformat"></a>longFormat</h2>

    

<p>The longFormat method takes data from the <code><a href='ExperimentList.html'>ExperimentList</a></code>
in a <code><a href='MultiAssayExperiment.html'>MultiAssayExperiment</a></code> and returns a uniform
<code><a href='https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html'>DataFrame</a></code>. The resulting DataFrame has columns indicating
primary, rowname, colname and value. This method can optionally include
colData columns with the <code>colDataCols</code> argument
(<code>MultiAssayExperiment</code> method only). The <code>i</code> argument
allows the user to specify the assay value in a
SummarizedExperiment. It directly relates to the <code>i</code>
argument in the assay method.</p>
    <h2 class="hasAnchor" id="wideformat"><a class="anchor" href="#wideformat"></a>wideFormat</h2>

    

<p>The <code>wideFormat</code> function returns standardized wide <a href='https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html'>DataFrame</a>
where each row represents a biological unit as in the <code>colData</code>.
Depending on the data and setup, biological units can be patients, tumors,
specimens, etc. Optionally, <code>colData</code> columns can be added to the
wide data output (see the <code>colDataCols</code> argument). Metadata columns are
generated based on the names produced in the wide format
DataFrame. These can be accessed via the
mcols function. See the <code>Arguments</code> and
<code>longFormat</code> sections for argument descriptions.</p>
    <h2 class="hasAnchor" id="hasrowranges"><a class="anchor" href="#hasrowranges"></a>hasRowRanges</h2>

    

<p>The <code>hasRowRanges</code> method identifies assays with associated ranged
row data by directly testing the method on the object. The result from the
test must be a GRanges class object to satisfy the test.</p>
    <h2 class="hasAnchor" id="getwithcoldata"><a class="anchor" href="#getwithcoldata"></a>getWithColData</h2>

    

<p>The <code>getWithColData</code> function allows the user to conveniently extract
a particular assay as indicated by the <strong><code>i</code></strong> index argument. It
will also attempt to provide the
<code>colData</code> along with the
extracted object using the <code>colData&lt;-</code> replacement
method when possible. Typically, this method is available for
SummarizedExperiment and <code>RaggedExperiment</code>
classes.</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Marcel Ramos, Levi Waldron.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


